### PR TITLE
Enhance concurrent access to callback in HandleManager

### PR DESCRIPTION
**Pull request by yuxihu**
1. Erase entry for the handle in callbacks map in ReleaseHandle
2. Retrieve callback under mutex and add NULL check in ExecuteCallback

### CODE REVIEW

File: horovod/mxnet/handle_manager.cc [hunk 1]
```diff
@@ -43,8 +43,16 @@ bool HandleManager::PollHandle(int handle) {
 }
 
 void HandleManager::ExecuteCallback(int handle) {
-  Callback cb = *callbacks_[handle];
-  cb();
+  std::unique_lock<std::mutex> lock(mutex_);
```
Comment (apeforest) [inline horovod/mxnet/handle_manager.cc]: Maybe using a lock_guard to lock and unlock this mutex.

File: horovod/mxnet/handle_manager.cc [hunk 2]
```diff
@@ -43,8 +43,16 @@ bool HandleManager::PollHandle(int handle) {
...
+  if (callbacks_.find(handle) == callbacks_.end()) {
+    return;
```
Comment (apeforest) [inline horovod/mxnet/handle_manager.cc]: lock_guard is needed, otherwise, the mutex_ is not unlocked.

File: horovod/mxnet/handle_manager.cc [hunk 3]
```diff
@@ -43,8 +43,16 @@ bool HandleManager::PollHandle(int handle) {
...
+  }
+  auto cb_ptr = callbacks_[handle];
```
Comment (apeforest) [inline horovod/mxnet/handle_manager.cc]: I would actually prefer the Callback data type here to increase readability

File: horovod/mxnet/handle_manager.cc (hunk 1 above)
Comment (yuxihu) [reply horovod/mxnet/handle_manager.cc]: See Line 51 for unlocking. Once we retrieved the callback object, we can release it earlier.

File: horovod/mxnet/handle_manager.cc (hunk 2 above)
Comment (yuxihu) [reply horovod/mxnet/handle_manager.cc]: unique_lock also automatically release the mutex when it goes out of scope due to exception, etc.

File: horovod/mxnet/handle_manager.cc (hunk 3 above)
Comment (yuxihu) [reply horovod/mxnet/handle_manager.cc]: It is easier to do nullptr check in this way.

File: horovod/mxnet/handle_manager.cc (hunk 2 above)
Comment (apeforest) [reply horovod/mxnet/handle_manager.cc]: never mind. as we discussed offline, unique_lock can also help to manage lock release.

File: horovod/mxnet/handle_manager.cc [hunk 4]
```diff
@@ -55,6 +63,7 @@ std::shared_ptr<Status> HandleManager::ReleaseHandle(int handle) {
   }
   auto status = results_[handle];
   results_.erase(handle);
+  callbacks_.erase(handle);
```
Comment (apeforest) [inline horovod/mxnet/handle_manager.cc]: Nice catch! I think this could be the root cause of random segmentation fault.

**Pull request merged**